diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt

```diff
@@ -107,56 +125,127 @@ class FileStructureRule(private val configRules: List<RulesConfig>) : Rule("file
             }
         }
     }
 
-    private fun rearrangeImports(node: ASTNode, imports: List<ASTNode>, sortedImports: List<ASTNode>) {
+    private fun rearrangeImports(node: ASTNode, imports: List<ASTNode>, sortedImportsGroups: List<List<ASTNode>>) {
         require(node.elementType == IMPORT_LIST)
         // move all commented lines among import before imports block
         node.getChildren(TokenSet.create(EOL_COMMENT))
-                .forEach {
-                    node.treeParent.addChild(it.clone() as ASTNode, node)
-                    node.treeParent.addChild(PsiWhiteSpaceImpl("\n"), node)
-                }
+            .forEach {
+                node.treeParent.addChild(it.clone() as ASTNode, node)
+                node.treeParent.addChild(PsiWhiteSpaceImpl("\n"), node)
+            }
 
         node.removeRange(imports.first(), imports.last())
-        sortedImports.forEachIndexed { index, importNode ->
-            if (index != 0) {
-                node.addChild(PsiWhiteSpaceImpl("\n"), null)
+        sortedImportsGroups.filterNot { it.isEmpty() }
+            .run {
+                forEachIndexed { groupIndex, group ->
+                    group.forEachIndexed { index, importNode ->
+                        node.addChild(importNode, null)
+                        if (index != group.size - 1) {
+                            node.addChild(PsiWhiteSpaceImpl("\n"), null)
+                        }
+                    }
+                    if (groupIndex != size - 1) {
+                        node.addChild(PsiWhiteSpaceImpl("\n\n"), null)
+                    }
+                }
             }
-            node.addChild(importNode, null)
-        }
     }
 
-    private fun ASTNode.getSiblingBlocks(copyrightComment: ASTNode?,
-                                         headerKdoc: ASTNode?,
-                                         fileAnnotations: ASTNode?,
-                                         packageDirectiveNode: ASTNode
+    private fun ASTNode.getSiblingBlocks(
+        copyrightComment: ASTNode?,
+        headerKdoc: ASTNode?,
+        fileAnnotations: ASTNode?,
+        packageDirectiveNode: ASTNode
     ): Pair<ASTNode?, ASTNode> = when (elementType) {
         BLOCK_COMMENT -> null to listOfNotNull(headerKdoc, fileAnnotations, packageDirectiveNode).first()
         KDOC -> copyrightComment to (fileAnnotations ?: packageDirectiveNode)
         FILE_ANNOTATION_LIST -> (headerKdoc ?: copyrightComment) to packageDirectiveNode
         else -> error("Only BLOCK_COMMENT, KDOC and FILE_ANNOTATION_LIST are valid inputs.")
     }
 
+    private data class ImportGroups(
+        val android: List<KtImportDirective>,
+        val ownDomain: List<KtImportDirective>,
+        val others: List<KtImportDirective>,
+        val javaCore: List<KtImportDirective>,
+        val kotlinCore: List<KtImportDirective>
+    ) {
+        fun toList() = listOf(android, ownDomain, others, javaCore, kotlinCore)
+
+        companion object {
```

Comment:
oh, I do not like these calculations in companion object. Why it should be here?

Reply:
I wanted to place a factory method inside the class itself. I sure can extract it to the Rule class level, but why is placing it in `companion object` bad?